master/buildbot/steps/source/mtn.py

```diff
@@ -216,52 +220,39 @@ def removeFiles(self, files):
                 return
         defer.returnValue(0)
 
-    def _clone(self, abandonOnFailure=False):
-        command = ['mtn', 'db', 'init', '--db', self.database]
-        d = self._dovccmd(command, abandonOnFailure=abandonOnFailure)
-        d.addCallback(lambda _: self._pull())
-        return d
-
     def _checkout(self, abandonOnFailure=False):
-        command = ['mtn', 'checkout', '.', '--db=%s' % (self.database)]
+        command = ['mtn', 'checkout', self.workdir, '--db', self.database]
         if self.revision:
             command.extend(['--revision', self.revision])
         command.extend(['--branch', self.branch])
+        return self._dovccmd('', command, abandonOnFailure=abandonOnFailure)
 
-        return self._dovccmd(command, abandonOnFailure=abandonOnFailure)
-
-    def _update(self, dopull=True, abandonOnFailure=False):
-        if dopull:
-            d = self._pull()
-        else:
-            d = defer.succeed(0)
-        command = ['mtn', 'update', '--db=%s' % (self.database)]
+    def _update(self, abandonOnFailure=False):
+        command = ['mtn', 'update']
         if self.revision:
             command.extend(['--revision', self.revision])
         else:
-            command.extend(["-r", "h:" + self.branch])
-        command.extend(["-b", self.branch])
-
-        d.addCallback(lambda _: self._dovccmd(command, abandonOnFailure=abandonOnFailure))
-        return d
+            command.extend(['--revision', 'h:' + self.branch])
+        command.extend(['--branch', self.branch])
+        return self._dovccmd(self.workdir, command,
+                             abandonOnFailure=abandonOnFailure)
 
     def _pull(self, abandonOnFailure=False):
-        command = ['mtn', 'pull', self.sourcedata,
-                   '--db=%s' % (self.database)]
+        command = ['mtn', 'pull', self.sourcedata, '--db', self.database]
         if self.progress:
             command.extend(['--ticker=dot'])
         else:
             command.extend(['--ticker=none'])
-        d = self._dovccmd(command, abandonOnFailure=abandonOnFailure)
+        d = self._dovccmd('', command, abandonOnFailure=abandonOnFailure)
```

Comment:
On 06/03/2015 07:11 AM, Mikhail Sobolev wrote:
> What I meant is: why not make |workdir| a keyword parameter with 
> default value to |None|? Calling |_dovccmd| with |''| looks a bit 
> strange. 
It's not None, it's an empty string, which is a valid relative path. Do 
you prefer '.'? That might clarify it.

WRT a default: I see no reason to prefer '.' over self.workdir as a 
default value (or vice versa), but prefer to keep it as an explicit 
argument I have to think about for every invocation of _dovccmd. I think 
it makes the code clearer to read. (Note that before, self.workdir was 
the default, as the mtn process was only ever executed in that directory.)

Regards

Markus Wanner